File: scanner.py
"""
scanner.py — Módulo de extracción de hardware.

Detecta CPU, RAM, GPU (NVIDIA) y almacenamiento del equipo local.
Maneja de forma segura la ausencia de GPU NVIDIA o drivers faltantes.
"""

from __future__ import annotations

import platform
from typing import Any

import psutil

try:
    import cpuinfo
except ImportError:
    cpuinfo = None  # type: ignore[assignment]


def _scan_cpu() -> dict[str, Any]:
    """Obtiene información del procesador."""
    info: dict[str, Any] = {
        "brand": "Desconocido",
        "physical_cores": psutil.cpu_count(logical=False) or 1,
        "logical_cores": psutil.cpu_count(logical=True) or 1,
        "frequency_mhz": 0.0,
        "arch": platform.machine(),
    }

    # Frecuencia vía psutil
    freq = psutil.cpu_freq()
    if freq:
        info["frequency_mhz"] = round(freq.max or freq.current, 1)

    # Marca/modelo vía py-cpuinfo (puede ser lento, ~1-2 s)
    if cpuinfo is not None:
        try:
            ci = cpuinfo.get_cpu_info()
            info["brand"] = ci.get("brand_raw", info["brand"])
            # py-cpuinfo reporta logical count; preferimos psutil para physical
        except Exception:
            pass  # Si falla cpuinfo, usamos lo de psutil

    return info


def _scan_ram() -> dict[str, Any]:
    """Obtiene información de la memoria RAM."""
    vm = psutil.virtual_memory()
    return {
        "total_gb": round(vm.total / (1024 ** 3), 1),
        "available_gb": round(vm.available / (1024 ** 3), 1),
        "percent_used": vm.percent,
    }


def _scan_gpu() -> dict[str, Any]:
    """
    Intenta detectar una GPU NVIDIA vía pynvml.

    Si no hay GPU NVIDIA, los drivers no están instalados, o pynvml
    no está disponible, devuelve un dict con available=False y valores en 0.
    """
    gpu_info: dict[str, Any] = {
        "available": False,
        "name": "No detectada",
        "vram_total_gb": 0.0,
        "vram_used_gb": 0.0,
        "vram_free_gb": 0.0,
        "driver_version": "N/A",
    }

    try:
        from pynvml import (
            nvmlInit,
            nvmlShutdown,
            nvmlDeviceGetHandleByIndex,
            nvmlDeviceGetName,
            nvmlDeviceGetMemoryInfo,
            nvmlSystemGetDriverVersion,
        )

        nvmlInit()
        try:
            handle = nvmlDeviceGetHandleByIndex(0)
            mem = nvmlDeviceGetMemoryInfo(handle)
            name_raw = nvmlDeviceGetName(handle)
            driver_raw = nvmlSystemGetDriverVersion()

            # pynvml puede devolver bytes o str según la versión
            name = name_raw.decode("utf-8") if isinstance(name_raw, bytes) else name_raw
            driver = driver_raw.decode("utf-8") if isinstance(driver_raw, bytes) else driver_raw

            gpu_info.update({
                "available": True,
                "name": name,
                "vram_total_gb": round(mem.total / (1024 ** 3), 1),
                "vram_used_gb": round(mem.used / (1024 ** 3), 1),
                "vram_free_gb": round(mem.free / (1024 ** 3), 1),
                "driver_version": driver,
            })
        finally:
            nvmlShutdown()

    except Exception:
        # Cualquier error (ImportError, NVMLError, etc.) → sin GPU
        pass

    return gpu_info


def _scan_storage() -> dict[str, Any]:
    """Obtiene información del disco principal (punto de montaje /)."""
    try:
        usage = psutil.disk_usage("/")
    except OSError:
        # Fallback para Windows
        usage = psutil.disk_usage("C:\\")

    return {
        "total_gb": round(usage.total / (1024 ** 3), 1),
        "used_gb": round(usage.used / (1024 ** 3), 1),
        "free_gb": round(usage.free / (1024 ** 3), 1),
        "percent_used": usage.percent,
    }


def scan_hardware() -> dict[str, Any]:
    """
    Ejecuta un escaneo completo del hardware y devuelve un diccionario
    consolidado con las secciones: cpu, ram, gpu, storage.
    """
    return {
        "cpu": _scan_cpu(),
        "ram": _scan_ram(),
        "gpu": _scan_gpu(),
        "storage": _scan_storage(),
    }


# Permite ejecutar el escáner de forma aislada para debug
if __name__ == "__main__":
    import json
    hw = scan_hardware()
    print(json.dumps(hw, indent=2, ensure_ascii=False))

File: arq_rust/src/scanner.rs
// ─────────────────────────────────────────────────────────────────────────────
// scanner.rs — Módulo de extracción de hardware.
//
// Detecta CPU, RAM, GPU (NVIDIA) y almacenamiento del equipo local.
// Maneja de forma segura la ausencia de GPU NVIDIA o drivers faltantes.
// ─────────────────────────────────────────────────────────────────────────────

use sysinfo::{System, Disks};

/// Información del procesador.
#[derive(Debug)]
pub struct CpuInfo {
    pub brand: String,
    pub physical_cores: usize,
    pub logical_cores: usize,
    pub frequency_mhz: u64,
    pub arch: String,
}

/// Información de la memoria RAM.
#[derive(Debug)]
pub struct RamInfo {
    pub total_gb: f64,
    pub available_gb: f64,
    pub percent_used: f64,
}

/// Información de la GPU NVIDIA.
#[derive(Debug)]
pub struct GpuInfo {
    pub name: String,
    pub vram_total_gb: f64,
    pub vram_used_gb: f64,
    pub vram_free_gb: f64,
    pub driver_version: String,
}

/// Información del almacenamiento principal.
#[derive(Debug)]
pub struct StorageInfo {
    pub total_gb: f64,
    pub used_gb: f64,
    pub free_gb: f64,
    pub percent_used: f64,
}

/// Contenedor con todo el hardware detectado.
#[derive(Debug)]
pub struct HardwareInfo {
    pub cpu: CpuInfo,
    pub ram: RamInfo,
    pub gpu: Option<GpuInfo>,
    pub storage: StorageInfo,
}

impl HardwareInfo {
    /// Devuelve el valor actual para un componente por nombre.
    pub fn actual_value(&self, component: &str) -> f64 {
        match component {
            "ram" => self.ram.total_gb,
            "vram" => self.gpu.as_ref().map_or(0.0, |g| g.vram_total_gb),
            "cores" => self.cpu.physical_cores as f64,
            "storage" => self.storage.free_gb,
            _ => 0.0,
        }
    }
}

/// Escanea la CPU usando sysinfo.
fn scan_cpu(sys: &System) -> CpuInfo {
    let cpus = sys.cpus();
    let brand = cpus
        .first()
        .map(|c| c.brand().to_string())
        .unwrap_or_else(|| "Desconocido".to_string());

    let frequency_mhz = cpus
        .first()
        .map(|c| c.frequency())
        .unwrap_or(0);

    CpuInfo {
        brand,
        physical_cores: sys.physical_core_count().unwrap_or(1),
        logical_cores: cpus.len().max(1),
        frequency_mhz,
        arch: std::env::consts::ARCH.to_string(),
    }
}

/// Escanea la memoria RAM usando sysinfo.
fn scan_ram(sys: &System) -> RamInfo {
    let total = sys.total_memory() as f64;
    let available = sys.available_memory() as f64;
    let used = sys.used_memory() as f64;

    let total_gb = total / (1024.0 * 1024.0 * 1024.0);
    let available_gb = available / (1024.0 * 1024.0 * 1024.0);
    let percent_used = if total > 0.0 {
        (used / total) * 100.0
    } else {
        0.0
    };

    RamInfo {
        total_gb: (total_gb * 10.0).round() / 10.0,
        available_gb: (available_gb * 10.0).round() / 10.0,
        percent_used: (percent_used * 10.0).round() / 10.0,
    }
}

/// Intenta detectar una GPU NVIDIA vía nvml-wrapper.
/// Devuelve None si no hay GPU NVIDIA o si fallan los drivers.
fn scan_gpu() -> Option<GpuInfo> {
    use nvml_wrapper::Nvml;

    let nvml = match Nvml::init() {
        Ok(n) => n,
        Err(_) => return None,
    };

    let device = match nvml.device_by_index(0) {
        Ok(d) => d,
        Err(_) => return None,
    };

    let name = device.name().unwrap_or_else(|_| "NVIDIA GPU".to_string());

    let mem = match device.memory_info() {
        Ok(m) => m,
        Err(_) => return None,
    };

    let driver = nvml
        .sys_driver_version()
        .unwrap_or_else(|_| "N/A".to_string());

    let total_gb = mem.total as f64 / (1024.0 * 1024.0 * 1024.0);
    let used_gb = mem.used as f64 / (1024.0 * 1024.0 * 1024.0);
    let free_gb = mem.free as f64 / (1024.0 * 1024.0 * 1024.0);

    Some(GpuInfo {
        name,
        vram_total_gb: (total_gb * 10.0).round() / 10.0,
        vram_used_gb: (used_gb * 10.0).round() / 10.0,
        vram_free_gb: (free_gb * 10.0).round() / 10.0,
        driver_version: driver,
    })
}

/// Escanea el almacenamiento principal.
fn scan_storage() -> StorageInfo {
    let disks = Disks::new_with_refreshed_list();

    // Buscar el disco raíz (/) o el de mayor tamaño
    let root_disk = disks
        .iter()
        .find(|d| d.mount_point() == std::path::Path::new("/"))
        .or_else(|| disks.iter().max_by_key(|d| d.total_space()));

    match root_disk {
        Some(disk) => {
            let total = disk.total_space() as f64;
            let free = disk.available_space() as f64;
            let used = total - free;

            let total_gb = total / (1024.0 * 1024.0 * 1024.0);
            let free_gb = free / (1024.0 * 1024.0 * 1024.0);
            let used_gb = used / (1024.0 * 1024.0 * 1024.0);
            let percent = if total > 0.0 {
                (used / total) * 100.0
            } else {
                0.0
            };

            StorageInfo {
                total_gb: (total_gb * 10.0).round() / 10.0,
                used_gb: (used_gb * 10.0).round() / 10.0,
                free_gb: (free_gb * 10.0).round() / 10.0,
                percent_used: (percent * 10.0).round() / 10.0,
            }
        }
        None => StorageInfo {
            total_gb: 0.0,
            used_gb: 0.0,
            free_gb: 0.0,
            percent_used: 0.0,
        },
    }
}

/// Ejecuta un escaneo completo del hardware del sistema.
pub fn scan_hardware() -> HardwareInfo {
    let mut sys = System::new_all();
    sys.refresh_all();

    HardwareInfo {
        cpu: scan_cpu(&sys),
        ram: scan_ram(&sys),
        gpu: scan_gpu(),
        storage: scan_storage(),
    }
}
